lib/matplotlib/dates.py

```diff
@@ -455,7 +457,7 @@ def date2num(d):
         if tzi is not None:
             # make datetime naive:
             d = [dt.astimezone(UTC).replace(tzinfo=None) for dt in d]
-            d = np.asarray(d)
+            d = np.ma.asarray(d) if masked else np.asarray(d)
         d = d.astype('datetime64[us]')
```

Comment:
I have a few concerns with the shadow typing for `d`. Note, this is just a suggestion - can think of better names.

```python
    # Unpack in case of e.g. Pandas or xarray object
    dnp = cbook._unpack_to_numpy(d)

    # make an iterable, but save state to unpack later:
    iterable = np.iterable(dnp)
    if not iterable:
        dnp = [dnp]

    masked = np.ma.is_masked(dnp)
    dnp = np.ma.asanyarray(dnp)

    # convert to datetime64 arrays, if not already:
    if not np.issubdtype(dnp.dtype, np.datetime64):
        # datetime arrays
        if not dnp.size:
            # deals with an empty array...
            return dnp
        tzi = getattr(dnp[0], 'tzinfo', None)
        if tzi is not None:
            # make datetime naive:
            d = [dt.astimezone(UTC).replace(tzinfo=None) for dt in dnp]
            d = np.ma.asarray(d, mask=dnp.get_mask()) if masked else np.asarray(dnp)
        d = d.astype('datetime64[us]')
```

Note, we should set the mask in the `tzi` changed array too, I believe?

Reply:
Ahh, obviously! 

Clearly that is not tested...